libs/experimental/langchain_experimental/llms/ollama_functions.py

```diff
@@ -354,6 +357,86 @@ def _generate(
             generations=[ChatGeneration(message=response_message_with_functions)]
         )
 
+    async def _agenerate(
+        self,
+        messages: List[BaseMessage],
+        stop: Optional[List[str]] = None,
+        run_manager: Optional[AsyncCallbackManagerForLLMRun] = None,
+        **kwargs: Any,
+    ) -> ChatResult:
+        functions = kwargs.get("functions", [])
+        if "functions" in kwargs:
+            del kwargs["functions"]
+        if "function_call" in kwargs:
+            functions = [
+                fn for fn in functions if fn["name"] == kwargs["function_call"]["name"]
+            ]
+            if not functions:
+                raise ValueError(
+                    "If `function_call` is specified, you must also pass a "
+                    "matching function in `functions`."
+                )
+            del kwargs["function_call"]
+        elif not functions:
+            functions.append(DEFAULT_RESPONSE_FUNCTION)
+        if _is_pydantic_class(functions[0]):
+            functions = [convert_to_ollama_tool(fn) for fn in functions]
+        system_message_prompt_template = SystemMessagePromptTemplate.from_template(
+            self.tool_system_prompt_template
+        )
+        system_message = system_message_prompt_template.format(
+            tools=json.dumps(functions, indent=2)
+        )
+        response_message = await super()._agenerate(
```

Comment:
@eyurtsev this is the difference, I am calling the async generate method which solves the serialization issue.

Reply:
Also I tested the async method on my machine and it solves the issue: #2222

Reply:
You can try it on your machine as well!!!